master/internal/api_experiment.go

```diff
@@ -2337,3 +2338,127 @@ func (a *apiServer) CreateTrial(
 
 	return resp, nil
 }
+
+func (a *apiServer) PutExperimentLabel(ctx context.Context,
+	req *apiv1.PutExperimentLabelRequest,
+) (*apiv1.PutExperimentLabelResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to get the user: %s", err)
+	}
+
+	tx, err := db.Bun().BeginTx(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err = tx.Rollback(); err != nil && err != sql.ErrTxDone {
+			logrus.WithError(err).Error("error rolling back transaction in create workspace")
+		}
+	}()
+
+	exp := &experimentv1.Experiment{}
+	query := db.Bun().NewSelect().
+		ModelTableExpr("experiments as e").
+		Model(exp).
+		Apply(getExperimentColumns).
+		Where("e.id = ?", req.ExperimentId).
```

Comment:
i think we need to `SELECT ... FOR UPDATE` here or we have the same problem on an even smaller level.

Reply:
at least that blocks other `SELECT ... FOR UPDATE`s on the same row, whereas if you to `begin transaction isolation level serializable` we have to retry or return 409 when we get a `could not serialize access due to concurrent update`.

Reply:
actually. i think `begin transaction isolation level serializable` and retries are probably the way to go. i think i prefer that over the issues `for update` can bring. curious what @ioga thinks too?

Reply:
I believe `SELECT FOR UPDATE` and then `UPDATE` is sufficient. that's the pattern I've been using elsewhere.

- why would we need serializable transactions? if any other reader reads this row, they'll either see the old version, or the new one, which is fine. if another writer was updating it, they need to do `select for update` themselves, which will block until this transaction.
- not sure what would cause retries here.

Reply:
> why would we need serializable transactions?

you can do just `SELECT` without `FOR UPDATE` then `UPDATE` with `isolation level serializable` and then one of the transactions will get `could not serialize access due to concurrent update` and need to retry, instead of blocking with `FOR UPDATE`. I guess `FOR UPDATE` is probably a better API experience.

Reply:
But that way you can let pg be smart rather than devs with `FOR UPDATE`.

Reply:
(only one of the txs gets aborted due to deadlock, though, and the other is free to finish, so ultimately the approaches are kinda the same)

Reply:
>I guess FOR UPDATE is probably a better API experience.

I agree, adding `.For("UPDATE")` is much simpler than having to wrap all transactions in retries.

>just because you can deadlock with FOR UPDATE

but that can't happen for this case with a single row update, correct?

Reply:
> but that can't happen for this case with a single row update, correct?

yeah, cannot.

Reply:
added a FOR UPDATE.